docs/src/developer_docs/model_interface.md

```diff
@@ -0,0 +1,227 @@
+
+- `initialize!(model::AbstractModel)`: called exactly once per `run!` before the
+  first time step.
+
+### Optional integrations
+
+While not required for Simulation itself, the following methods enable the rest
+of the Oceananigans ecosystem to “see” the model:
+
+- `fields(model)` and `prognostic_fields(model)`: return `NamedTuple`s of
+  fields so diagnostics, output writers, and NaN checkers know what to touch.
+
+- `default_nan_checker(model)`: customize the `NaNChecker` that Simulation adds
+  by default.
+
+## Models that implement this interface
+
+Several models across the CliMA ecosystem implement this interface:
+
+**Oceananigans.jl**
+- `NonhydrostaticModel`: solves the incompressible Navier-Stokes equations
+- `HydrostaticFreeSurfaceModel`: solves the hydrostatic Boussinesq equations with a free surface
+- `ShallowWaterModel`: solves the shallow water equations
+
+**ClimaOcean.jl**
+- `OceanSeaIceModel`: couples ocean, sea ice, and atmosphere components for Earth system modeling.
+                      The components themselves may be `Simulation` that contain `AbstractModel`,
+                      generating a nested structure.
+
+**ClimaSeaIce.jl**
+- `SeaIceModel`: simulates sea ice thermodynamics and dynamics
+
+**Breeze.jl**
+- `AtmosphereModel`: simulates atmospheric dynamics
+
+## Example: a zero-dimensional `LorenzModel`
+
+The example below shows a deliberately tiny model that integrates the Lorenz
+system on a 0D grid. The implementation demonstrates how little is required:
+store a `grid` and `clock`, provide `time_step!` and `update_state!`
+implementations, and rely on fallbacks for the rest. Here we use an explicit
+forward-Euler step so all of the logic lives directly inside `time_step!`.
+
+### Implementing the interface
+
+```@example model_interface
+using Oceananigans
+using Oceananigans.Models: AbstractModel
+using Oceananigans.Simulations: Simulation, run!
+using Oceananigans.TimeSteppers: Clock, tick!
+import Oceananigans.TimeSteppers: update_state!, time_step!
+import Oceananigans.Fields: set!
+using Oceananigans: TendencyCallsite, UpdateStateCallsite
+
+mutable struct LorenzModel{G, C, P, S} <: AbstractModel{Nothing, Nothing}
+    grid :: G
+    clock :: C
+    parameters :: P
+    state :: S
+end
+
+function LorenzModel(FT=Oceananigans.defaults.FloatType;
+                     σ = 10, ρ = 28, β = 8/3)
+
+    grid = RectilinearGrid(size=(), topology=(Flat, Flat, Flat))
+    clock = Clock{FT}(time = zero(FT))
+    parameters = (σ = FT(σ), ρ = FT(ρ), β = FT(β))
+    state = (x = CenterField(grid), y = CenterField(grid), z = CenterField(grid))
+
+    return LorenzModel(grid, clock, parameters, state)
+end
+
+Base.summary(::LorenzModel) = "LorenzModel"
+
+function set!(model::LorenzModel; kw...)
+    for (name, val) in kw
+        set!(model.state[name], val)
+    end
+    return nothing
+end
+
+function update_state!(model::LorenzModel, callbacks = []; compute_tendencies=true)
+    for callback in callbacks
```

Comment:
is this loop always present? could we define a wrapper dispatch for `update_state!` that relieves the downstream implemetnations of always needing to write this?

Reply:
That is possible but outside the scope of this PR because it would be a significant refactor for all models. But note that this form for `update_state!` is actually only needed internally within a model --- because it is called by `time_step!(model, dt)`. I believe when `update_state!` is called by simulation, there is no second argument.